In [2]:
import pandas as pd

In [3]:
data = pd.read_excel('/content/data_1.2.xlsx')

In [4]:
data

,КЛАСС,ПОЛ,ИНТЕЛЛЕКТ,МОТИВАЦИЯ,СРЕДНИЙ_БАЛЛ,Unnamed: 5,Unnamed: 6,Unnamed: 7
0,6,2.0,16,24,4.71,NaN,0.21155,H1
1,6,2.0,28,19,3.71,NaN,NaN,NaN
2,6,1.0,25,26,3.86,NaN,NaN,NaN
3,6,1.0,35,23,4.14,NaN,NaN,NaN
4,6,1.0,40,20,4.29,NaN,NaN,NaN
...,...,...,...,...,...,...,...,...
665,10,2.0,44,23,3.40,NaN,NaN,NaN
666,10,1.0,40,24,4.10,NaN,NaN,NaN
667,10,1.0,33,23,3.30,NaN,NaN,NaN
668,10,1.0,36,13,3.10,NaN,NaN,NaN


# Task
Perform Mann-Whitney U tests on the 'ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', and 'СРЕДНИЙ_БАЛЛ' columns for all unique pairs of classes from the 'КЛАСС' column in the `data` DataFrame, and then present the U statistics and p-values in a table.

## Prepare the data for analysis

### Subtask:
Ensure necessary libraries are imported and the data is ready for statistical tests.


**Reasoning**:
First, I will import the `mannwhitneyu` function from `scipy.stats` as requested, which is necessary for potential statistical tests.



In [5]:
from scipy.stats import mannwhitneyu
print("mannwhitneyu imported successfully")

mannwhitneyu imported successfully


**Reasoning**:
Now that `mannwhitneyu` is imported, I will check the data types of the relevant columns in the `data` DataFrame to ensure they are suitable for numerical analysis, as requested.



In [11]:
print(data[['ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ', 'КЛАСС']].info())

<class 'pandas.core.frame.DataFrame'>
Index: 660 entries, 0 to 669
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ИНТЕЛЛЕКТ     660 non-null    int64  
 1   МОТИВАЦИЯ     660 non-null    int64  
 2   СРЕДНИЙ_БАЛЛ  660 non-null    float64
 3   КЛАСС         660 non-null    int64  
dtypes: float64(1), int64(3)
memory usage: 25.8 KB
None


**Reasoning**:
The previous error indicated that the column names 'ИНТЕЛЛЕКТ' and 'МОТИВАЦИЯ' were not found in the DataFrame. This can happen due to subtle differences in characters or hidden whitespaces. To identify the exact column names, I will print all column names of the `data` DataFrame.



In [7]:
print(data.columns.tolist())

['КЛАСС', 'ПОЛ', ' ИНТЕЛЛЕКТ', ' МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ', 'Unnamed: 5', 'Unnamed: 6', 'Unnamed: 7']


**Reasoning**:
The previous output showed that the columns ' ИНТЕЛЛЕКТ' and ' МОТИВАЦИЯ' have leading spaces. To ensure proper data manipulation and analysis, I will rename these columns to remove the leading spaces. After renaming, I will proceed to check the data types of the relevant columns, including the corrected ones, to ensure they are suitable for numerical analysis, as specified in the subtask.



In [8]:
data = data.rename(columns={' ИНТЕЛЛЕКТ': 'ИНТЕЛЛЕКТ', ' МОТИВАЦИЯ': 'МОТИВАЦИЯ'})
print(data[['ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ', 'КЛАСС']].info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 670 entries, 0 to 669
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ИНТЕЛЛЕКТ     670 non-null    int64  
 1   МОТИВАЦИЯ     670 non-null    int64  
 2   СРЕДНИЙ_БАЛЛ  660 non-null    float64
 3   КЛАСС         670 non-null    int64  
dtypes: float64(1), int64(3)
memory usage: 21.1 KB
None


**Reasoning**:
The previous output indicated that the `СРЕДНИЙ_БАЛЛ` column has missing values (660 non-null out of 670 entries). To ensure accurate statistical analysis, I will drop rows with missing values in this column, as requested by the subtask to have data ready for statistical tests.



In [9]:
data.dropna(subset=['СРЕДНИЙ_БАЛЛ'], inplace=True)
print(data[['ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ', 'КЛАСС']].info())

<class 'pandas.core.frame.DataFrame'>
Index: 660 entries, 0 to 669
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   ИНТЕЛЛЕКТ     660 non-null    int64  
 1   МОТИВАЦИЯ     660 non-null    int64  
 2   СРЕДНИЙ_БАЛЛ  660 non-null    float64
 3   КЛАСС         660 non-null    int64  
dtypes: float64(1), int64(3)
memory usage: 25.8 KB
None


## Identify unique classes and generate pairs

### Subtask:
Extract unique values from the 'КЛАСС' column and create all possible unique pairs of these classes for comparison.


**Reasoning**:
To extract unique values from the 'КЛАСС' column and generate all possible unique pairs of these classes, I will first get the unique values from the 'КЛАСС' column of the `data` DataFrame. Then, I will import the `itertools` module, which contains the `combinations` function needed to generate all unique pairs of these classes.



In [10]:
import itertools

unique_classes = data['КЛАСС'].unique()
class_pairs = list(itertools.combinations(unique_classes, 2))

print(f"Unique Classes: {unique_classes}")
print(f"Class Pairs: {class_pairs}")

Unique Classes: [ 6  7  8  9 10]
Class Pairs: [(np.int64(6), np.int64(7)), (np.int64(6), np.int64(8)), (np.int64(6), np.int64(9)), (np.int64(6), np.int64(10)), (np.int64(7), np.int64(8)), (np.int64(7), np.int64(9)), (np.int64(7), np.int64(10)), (np.int64(8), np.int64(9)), (np.int64(8), np.int64(10)), (np.int64(9), np.int64(10))]


## Perform Mann-Whitney U test for each pair and variable

### Subtask:
Iterate through each pair of classes and for each specified variable ('ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ'), perform the Mann-Whitney U test. Collect the U statistic and p-value from each test.


**Reasoning**:
I will now iterate through each unique pair of classes and for each specified variable, perform the Mann-Whitney U test, collecting the U statistic and p-value as requested by the subtask instructions.



In [12]:
variables_to_test = ['ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ']
mw_results = []

for class1, class2 in class_pairs:
    class1_data = data[data['КЛАСС'] == class1]
    class2_data = data[data['КЛАСС'] == class2]

    for variable in variables_to_test:
        # Perform Mann-Whitney U test
        u_statistic, p_value = mannwhitneyu(class1_data[variable], class2_data[variable])

        # Store results
        mw_results.append({
            'Class 1': class1,
            'Class 2': class2,
            'Variable': variable,
            'U-statistic': u_statistic,
            'p-value': p_value
        })

print(f"Collected {len(mw_results)} test results.")

Collected 30 test results.


In [13]:
import pandas as pd

mw_results_df = pd.DataFrame(mw_results)
display(mw_results_df)

,Class 1,Class 2,Variable,U-statistic,p-value
0,6,7,ИНТЕЛЛЕКТ,6715.5,7.582910e-03
1,6,7,МОТИВАЦИЯ,9265.0,1.110438e-01
2,6,7,СРЕДНИЙ_БАЛЛ,10764.5,4.032608e-05
3,6,8,ИНТЕЛЛЕКТ,7734.5,2.511935e-02
4,6,8,МОТИВАЦИЯ,10649.0,2.218951e-02
5,6,8,СРЕДНИЙ_БАЛЛ,12855.0,1.131131e-08
6,6,9,ИНТЕЛЛЕКТ,5221.5,4.701925e-09
7,6,9,МОТИВАЦИЯ,9499.5,3.500258e-01
8,6,9,СРЕДНИЙ_БАЛЛ,11749.0,6.725442e-06
9,6,10,ИНТЕЛЛЕКТ,4724.0,1.941191e-10


In [14]:
significant_results_df = mw_results_df[mw_results_df['p-value'] < 0.05]
display(significant_results_df)

,Class 1,Class 2,Variable,U-statistic,p-value
0,6,7,ИНТЕЛЛЕКТ,6715.5,7.582910e-03
2,6,7,СРЕДНИЙ_БАЛЛ,10764.5,4.032608e-05
3,6,8,ИНТЕЛЛЕКТ,7734.5,2.511935e-02
4,6,8,МОТИВАЦИЯ,10649.0,2.218951e-02
5,6,8,СРЕДНИЙ_БАЛЛ,12855.0,1.131131e-08
6,6,9,ИНТЕЛЛЕКТ,5221.5,4.701925e-09
8,6,9,СРЕДНИЙ_БАЛЛ,11749.0,6.725442e-06
9,6,10,ИНТЕЛЛЕКТ,4724.0,1.941191e-10
11,6,10,СРЕДНИЙ_БАЛЛ,11474.5,4.352698e-06
15,7,9,ИНТЕЛЛЕКТ,6573.5,2.742771e-03


## Present results in a table

### Subtask:
Organize the collected U statistics and p-values into a readable table, clearly indicating the classes compared and the variable analyzed.


## Summary:

### Data Analysis Key Findings
*   The raw data initially had columns with leading spaces (' ИНТЕЛЛЕКТ', ' МОТИВАЦИЯ'), which were corrected by renaming them to 'ИНТЕЛЛЕКТ' and 'МОТИВАЦИЯ'.
*   Ten missing values were identified and removed from the 'СРЕДНИЙ_БАЛЛ' column, resulting in a clean dataset of 660 entries suitable for analysis.
*   Five unique classes (6, 7, 8, 9, 10) were identified from the 'КЛАСС' column, leading to 10 unique pairs for comparison.
*   A total of 30 Mann-Whitney U tests were successfully performed by iterating through these 10 class pairs and three variables ('ИНТЕЛЛЕКТ', 'МОТИВАЦИЯ', 'СРЕДНИЙ_БАЛЛ'), collecting the U statistic and p-value for each.

### Insights or Next Steps
*   The collected U statistics and p-values are ready to be organized and presented in a table for clearer interpretation of the differences between classes across the selected variables.
*   The statistical significance of the differences between class pairs for each variable can now be evaluated by examining the p-values from the Mann-Whitney U tests.
